galley/pkg/runtime/processor.go

```diff
@@ -77,74 +72,57 @@ func newProcessor(
 		state:           state,
 		source:          src,
 		postProcessHook: postProcessHook,
-		done:            make(chan struct{}),
-		stopped:         make(chan struct{}),
+		worker:          util.NewWorker("runtime processor", scope),
+		eventCh:         make(chan resource.Event, 1024),
 		lastEventTime:   now,
 	}
 }
 
 // Start the processor. This will cause processor to listen to incoming events from the provider
 // and publish component configuration via the Distributor.
 func (p *Processor) Start() error {
-	scope.Info("Starting processor...")
+	return p.worker.Start(func(stopCh chan struct{}, stoppedCh chan struct{}) error {
+		scope.Info("Starting processor...")
+
+		err := p.source.Start(func(e resource.Event) {
+			p.eventCh <- e
+		})
+		if err != nil {
+			scope.Warnf("Unable to Start source: %v", err)
+			return err
+		}
 
-	if p.events != nil {
-		scope.Warn("Processor has already started")
-		return errors.New("already started")
-	}
+		go p.process(stopCh, stoppedCh)
 
-	events := make(chan resource.Event, 1024)
-	err := p.source.Start(func(e resource.Event) {
-		events <- e
+		return nil
 	})
-	if err != nil {
-		scope.Warnf("Unable to Start source: %v", err)
-		return err
-	}
-
-	p.events = events
-
-	go p.process()
-
-	return nil
 }
 
 // Stop the processor.
 func (p *Processor) Stop() {
 	scope.Info("Stopping processor...")
```

Comment:
This function is already blocking. As such, I don't think we need pre/post mode, which makes the worker logic complicated. It looks like you're trying to reuse the lock in worker here, which is subtle and hard to understand.

Instead this type should be responsible managing its own internal sate ordering. It should stop the worker, and close the event channel here, as part of synchronous function call.

Reply:
done.